File: python/medium/disjoint_set.py
"""
Disjoint-set Union/Find
"""
def disjointFind(data: list, i: int) -> int:
  if i != data[i]:
    data[i] = disjointFind(data, data[i])
  return data[i]

def union(data: list, i: int, j: int) -> None:
  pi, pj = disjointFind(data, i), disjointFind(data, j)
  if pi != pj:
    data[pi] = pj

def isConnected(data: list, i: int, j: int) -> bool:
  return disjointFind(data, i) == disjointFind(data, j)

File: java/src/data_structures/DisjointSet.java
package data_structures;

public class DisjointSet {

  private int[] data;

  public DisjointSet(int[] data) {
    this.data = data;
  }

  public int find(int i) {
    if (i != this.data[i]) {
      this.data[i] = this.find(this.data[i]);
    }
    return this.data[i];
  }

  public void union(int i, int j) {
    int pi = this.find(i);
    int pj = this.find(j);
    if (pi != pj) {
      this.data[pi]= this.data[pj];
    }
  }

  public boolean isConnected(int i, int j) {
    return this.find(i) == this.find(j);
  }

  public static void main(String[] args) {
    int[] nums = new int[] {0, 1, 2, 3, 4, 5, 6, 7, 8};
    DisjointSet disjointSet = new DisjointSet(nums);
    disjointSet.union(1, 8);
    System.out.println(disjointSet.find(1));
  }
}